native/vendor/google/billing/GoogleBillingHelper.cpp

```diff
@@ -42,555 +48,354 @@ namespace {
+    JniHelper::callStaticVoidMethod(JCLS_BILLING, "isAlternativeBillingOnlyAvailableAsync", tag, callbackId);
 }
 
-void GoogleBillingHelper::isExternalOfferAvailableAsync() {
-    JniHelper::callStaticVoidMethod(JCLS_BILLING, "isExternalOfferAvailableAsync");
+void GoogleBillingHelper::createExternalOfferReportingDetailsAsync(int tag, int callbackId) {
+    JniHelper::callStaticVoidMethod(JCLS_BILLING, "createExternalOfferReportingDetailsAsync", tag, callbackId);
 }
 
-BillingResult* GoogleBillingHelper::showAlternativeBillingOnlyInformationDialog() {
-    return callFunctionAndReturnBillingResult("showAlternativeBillingOnlyInformationDialog");
+void GoogleBillingHelper::isExternalOfferAvailableAsync(int tag, int callbackId) {
+    JniHelper::callStaticVoidMethod(JCLS_BILLING, "isExternalOfferAvailableAsync", tag, callbackId);
 }
 
-BillingResult* GoogleBillingHelper::showExternalOfferInformationDialog() {
-    return callFunctionAndReturnBillingResult("showExternalOfferInformationDialog");
+BillingResult* GoogleBillingHelper::showAlternativeBillingOnlyInformationDialog(int tag, int callbackId) {
+    return cc::JniBilling::callFunctionAndReturnBillingResult("showAlternativeBillingOnlyInformationDialog", tag, callbackId);
 }
 
-BillingResult* GoogleBillingHelper::showInAppMessages() {
-    return callFunctionAndReturnBillingResult("showInAppMessages");
+BillingResult* GoogleBillingHelper::showExternalOfferInformationDialog(int tag, int callbackId) {
+    return cc::JniBilling::callFunctionAndReturnBillingResult("showExternalOfferInformationDialog", tag, callbackId);
 }
 
-void GoogleBillingHelper::launchBillingFlow(const std::vector<ProductDetails*>& productDetailsList, const std::string& selectedOfferToken) {
-    if (productDetailsList.empty()) {
-        return;
-    }
+BillingResult* GoogleBillingHelper::showInAppMessages(int tag, int callbackId, const std::vector<int>& inAppMessageCategoryId) {
     auto* env = JniHelper::getEnv();
-    const int size = productDetailsList.size();
-    jintArray result = env->NewIntArray(size);
-    jint* buf = new jint[size];
-    for (int i = 0; i < size; ++i) {
-        buf[i] = productDetailsList[i]->_id;
-    }
-    env->SetIntArrayRegion(result, 0, size, buf);
-    delete[] buf;
     cc::JniMethodInfo t;
-    cc::JniHelper::getStaticMethodInfo(t, JCLS_BILLING, "launchBillingFlow", "([ILjava/lang/String;)V");
-    jstring offerToken = env->NewStringUTF(selectedOfferToken.c_str());
-    t.env->CallStaticVoidMethod(t.classID, t.methodID, result, offerToken);
+    if (cc::JniHelper::getStaticMethodInfo(t, JCLS_BILLING, "showInAppMessages", "(II[I)Lcom/android/billingclient/api/BillingResult;")) {
+        const int size = inAppMessageCategoryId.size();
+        jintArray result = env->NewIntArray(size);
+        jint* buf = new jint[size];
+        for (int i = 0; i < size; ++i) {
+            buf[i] = inAppMessageCategoryId[i];
+        }
+        env->SetIntArrayRegion(result, 0, size, buf);
```

Comment:
Could we just use the parameter `inAppMessageCategoryId` instead of re-allocate a new array to SetIntArrayRegion?
```suggestion
        env->SetIntArrayRegion(result, 0, inAppMessageCategoryId.size(), inAppMessageCategoryId.data());
```

Reply:
fixed.